pycbc/waveform/waveform_modes.py

```diff
@@ -0,0 +1,491 @@
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation; either version 3 of the License, or (at your
+# option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+"""Provides functions and utilities for generating waveforms mode-by-mode.
+"""
+
+from string import Formatter
+import lal
+import lalsimulation
+
+from pycbc.types import (TimeSeries, FrequencySeries)
+from .waveform import (props, _check_lal_pars, check_args)
+from . import parameters
+
+
+def _formatdocstr(docstr):
+    """Utility for formatting docstrings with parameter information.
+    """
+    return docstr.format(
+        **{_p[1]: getattr(parameters, _p[1]).docstr(
+            prefix="    ", include_label=False).lstrip(' ')
+           for _p in Formatter().parse(docstr) if _p[1] is not None
+           })
+
+
+def _formatdocstrlist(docstr, paramlist, skip_params=None):
+    """Utility for formatting docstrings with parameter information.
+    """
+    if skip_params is None:
+        skip_params = []
+    pl = '\n'.join([_p.docstr(prefix="    ", include_label=False)
+                    for _p in paramlist if _p not in skip_params]).lstrip(' ')
+    return docstr.format(params=pl)
+
+
+def sum_modes(hlms, inclination, phi):
+    """Applies spherical harmonics and sums modes to produce a plus and cross
+    polarization.
+
+    Parameters
+    ----------
+    hlms : dict
+        Dictionary of ``(l, m)`` -> complex ``hlm``. The ``hlm`` may be a
+        complex number or array, or complex ``TimeSeries``. All modes in the
+        dictionary will be summed.
+    inclination : float
+        The inclination to use.
+    phi : float
+        The phase to use.
+
+    Returns
+    -------
+    complex float or array
+        The plus and cross polarization as a complex number. The real part
+        gives the plus, the negative imaginary part the cross.
+    """
+    out = None
+    for mode in hlms:
+        ell, m = mode
+        hlm = hlms[ell, m]
+        ylm = lal.SpinWeightedSphericalHarmonic(inclination, phi, -2, ell, m)
+        if out is None:
+            out = ylm * hlm
+        else:
+            out += ylm * hlm
+    return out
+
+
+def default_modes(approximant):
+    """Returns the default modes for the given approximant.
+    """
+    # FIXME: this should be replaced to a call to a lalsimulation function,
+    # whenever that's added
+    if approximant in ['IMRPhenomXPHM', 'IMRPhenomXHM']:
+        # according to arXiv:2004.06503
+        ma = [(2, 2), (2, 1), (3, 3), (3, 2), (4, 4)]
```

Comment:
Are these the modes in the co-precessing frame, or in the radiation frame? I assume that the radiation frame is what is really important given what's above ... But then `XPHM` *must* have the other 2,X modes in the radiation frame or it's just broken!

(I think this does need to be in the docstrings, but interesting what is intended with precessing waveforms first.)

Reply:
Yeah ... Actually I think there is inconsistency here, as I think you have defined the Phenom modes in the co-precessing frame, but NRSur seems to be in radiation frame, as no NR waveform is modelling (4,0) mode [well]!

Reply:
They must be defined in the intertial frame (closely related to the radiation frame but remember there is not inclination at this point).

Reply:
I'll let Collin respond, but my understanding is that *only* the NRSur models (or rather two of them) actually support this output at the moment.

Reply:
Yeah, good point about radiation frame.

... I think that SpinTaylorTX waveforms also support this output, but that may not be relevant for you and could be added as/when needed if there's no consistent interface [yet] in lalsim.

Reply:
Yeah, at the moment there's no common interface. @Cyberface had mentioned that they might have something in the works, but don't know when that will be.

I haven't used SpinTaylor in a looonngg time...

Reply:
So it sounds like only `NRSur7dq4` is actually good to go right now? I think based on the above, this *definitely* cannot do the right thing for the Phenom precessing approximants. I would prefer these be removed for now (I can believe that the non-precessing Phenom approximants are okay).

Reply:
This block of code is just returning what the default modes are for each approximant. That is correct, and is generally useful, so leaving in.

Reply:
I agree that it's returning the default modes for `IMRPhenomXHM` in the normal coordinate system that we are using and that this is useful. However, it is *not* returning the default modes for `IMRPhenomXPHM` in that same coordinate system (surely) and I'm worried about the potential for confusion (or mistakes) further down the line related to this.